Added console input History, and trap Error in Try-Catch

mingfudesign opened the pull request:
Added $history to track latest 99 valid inputs
Added commands for showing console inputs
Added pre-amble at Console start to list available commands
Try-Catch traps Errors to prevent Console from exiting automatically

mingfudesign committed:
Added console input History and trap Error

Console tracks latest 99 valid inputs
Added commands for showing console inputs
Added pre-amble at Console start to list available commands
Try-Catch traps Errors to prevent Console from exiting automatically

WanWizard commented on classes/console.php:
```diff
@@ -2,12 +2,12 @@
  * @copyright  2010 - 2019 Fuel Development Team
- * @link       https://fuelphp.com
+ * @link	   https://fuelphp.com
  */
```
Please undo the changes in the header (no tabs).

WanWizard commented on classes/console.php:
```diff
@@ -64,6 +64,29 @@ public static function help()
 	}
 
+	private const MAX_HISTORY = 99;
+	private $history = [];
```
Class variables should be defined at the top, before the constructor, not in the middle of the code.
Also, never use "private", use "protected". All Fuel classes should be fully extendable.

WanWizard commented on classes/console.php:
```diff
@@ -64,6 +64,29 @@ public static function help()
+	private const MAX_HISTORY = 99;
+	private $history = [];
+
+	private function push_history($line)
```
protected instead of private.

WanWizard commented on classes/console.php:
```diff
@@ -64,6 +64,29 @@ public static function help()
+		$this->history = array_slice($this->history, -self::MAX_HISTORY);
+	}
+
+	private function pop_history()
```
protected instead of private.

WanWizard commented on classes/console.php:
```diff
@@ -64,6 +64,29 @@ public static function help()
+		array_pop($this->history);
+	}
+
+	private function show_history()
```
protected instead of private.

WanWizard commented on classes/console.php:
```diff
@@ -75,6 +98,8 @@ private function main()
 			PHP_OS
 		));
 
+		\Cli::write(['', 'Commands', ':q | quit - exit the console', ':h | history - show transcript', '']);
```
No short array notation please, Fuel still requires PHP 5.6 compatibility.

WanWizard commented on classes/console.php:
```diff
@@ -64,6 +64,29 @@ public static function help()
+		\Cli::write('');
+	}
+
 	private function main()
```
Please while you're at it, please change this to protected too...

WanWizard commented on classes/console.php:
```diff
@@ -88,13 +113,19 @@ private function main()
 			}
 
-			if ($__line == 'quit')
+			if ($__line == ':q' || $__line == 'quit')
```
Please use "or" instead of "||" ( coding standards )

WanWizard commented on classes/console.php:
```diff
@@ -88,13 +113,19 @@ private function main()
 			{
 				break;
 			}
+			elseif ($__line == ':h' || $__line == 'history')
```
Please use "or" instead of "||" ( coding standards )

WanWizard commented on classes/console.php:
```diff
@@ -113,8 +144,11 @@ private function main()
 				$ret = eval("unset(\$__line); $__line;");
 			}
-			catch(\Exception $e)
+			catch(\Exception|\Error $e)
```
Catching multiple exceptions isnt supported in PHP < 7.1, so please split this.

WanWizard reviewed (changes requested):
Thanks for your contribution, much appreciated!

Could you please address the comments?

sinclairday committed:
Clean up for submission to Fuel upstream

conform to coding standards; reverted unintended formatting change;
changed to protected for easier subclassing; PHP <5.6 compatibility

mingfudesign commented:
I think, with the latest changes, all of the comments have been addressed. Please check.

WanWizard commented:
Looks fine, thanks !
